server/jms/src/main/java/org/hawkular/apm/server/jms/RetryCapableMDB.java

```diff
@@ -95,8 +98,7 @@ public void onMessage(Message message) {
             process(tenantId, items, retryCount);
 
         } catch (Exception e) {
-            // TODO: Handle nak of JMS message?
```

Comment:
What is missing here, IMHO, is the handling of an exceptional case where you don't want the retry to happen. Right now, you force the processor implementations to throw `Error`s if they encounter a "grave condition" that should not result in a retry. IMHO, that is too strong.

IMHO something along the lines of `RetryAttemptException` could be thrown from (bulk)process and caught here. The processor impls then would have clear "guidance" on how to categorize the exceptions they encounter - if you want a retry to happen throw `RetryAttemptException`. If the error is severe or unexpected enough that a retry is not warranted, throw anything else.

But anyway - I don't know the codebase. If you think this suggestion is overly complicated for this part of the code and is not worth the effort, I am not against merging this PR as it is. It fulfills the purpose.